dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1124,6 +1124,25 @@ RtpsUdpDataLink::RtpsReader::process_data_i(const RTPS::DataSubmessage& data,
     WriterInfo& info = wi->second;
     SequenceNumber seq;
     seq.setValue(data.writerSN.high, data.writerSN.low);
+
+    // determines if this is a direct write to this reader
```

Comment:
As we discussed, this will be cleaner to integrate in RtpsUdpReceiveStrategy.  RtpsUdpDataLink has additional functions that are only used for reliable readers.

Reply:
I forgot to write it here yesterday, but this was attempting to be an answer to your point #3 from the large post above.  The purpose of RtpsUdpDataLink's internal data structures like RtpsWriter and ReaderInfo is for reliability (see comments in RtpsUdpDataLink.h).  When reliability isn't needed, the transport framework should support the functionality needed.

Reply:
Yeah, I understand a little better now. So the other checks like the duplicate seq# are not required by the specs for BestEffort. I factored out the DirectedWrite logic into RtpsUdpReceiveStrategy. The logic is comparing the inline qos reader id with the current reader id. How do we get the current reader id in RtpsUdpReceiveStrategy? To call it elsewhere, such as in RtpsUdpDataLink::RtpsReader, then we still need the readers to be inserted in RtpsUdpDataLink::associated(). I wish associated() were not tied to reliability settings.

Reply:
Duplicate sequence numbers probably need to be handled somewhere, but it's not in the scope of this effort.
The main point here is the division of responsibilities between the framework and the specific transport type.  The design goal is to have the framework handle things that are common across transports.  The framework already has support for DataLink::data_received and DataLink::data_received_include, so those should be used here.
So to try and answer your question: there is no "current reader id", that doesn't appear at this layer.  Instead the specific transport type (RtpsUdpReceiveStrategy) can make method calls on the DataLink to tell it to deliver this sample only to those associated readers that have a specific GUID (or are in given set of GUIDs).